shared-libs/rules-engine/src/pouchdb-provider.js

```diff
@@ -46,32 +43,41 @@ const medicPouchProvider = db => {
         })
     ),
 
+    stateChangeCallback: docUpdateClosure(db),
+
+    commitTargetDoc: (assign, userDoc, docTag) => {
+      const userContactId = userDoc && userDoc._id;
+      const _id = `target-${docTag}-${userContactId}`;
+      if (!commitTargetDocClosuers[_id]) {
+        const createNew = () => ({
+          _id,
+          type: 'target',
+          user: userContactId,
+        });
+
+        return db.get(_id)
+          .catch(createNew)
+          .then(existingDoc => {
+            const closure = docUpdateClosure(db);
+            commitTargetDocClosuers[_id] = doc => closure(existingDoc, doc);
+            return commitTargetDocClosuers[_id](assign);
+          });
+      }
+
+      return commitTargetDocClosuers[_id](assign);
+    },
```

Comment:
I saw conflict errors a few times during testing and decided to add this. 

There are several things which might result in either the `rulesStateStore` or this target doc being written twice quickly: users clicking quickly on different tabs, the "freshness" thread in the rules-engine service firing at the same time the user clicks the target tab, or the changes feed in the tasks controller firing at the same time the user clicks the tasks/targets tab. None of these scenarios are particularly likely. Avoiding conflicts in the state store is important, because if it gets out of sync then users see incorrect thigns. But conflicts in the target document isn't that important - just stale analytics data. So one could reasonable argue this is overkill and should just be removed.

Reply:
I think we should update the doc daily instead of on each view - which would also solve the conflicts for target docs.

Reply:
> I think we should update the doc daily instead of on each view - which would also solve the conflicts for target docs.

If I understand correctly this would mean if someone completed a task then the task completed count in the target wouldn't increment until midnight, right? We might be able to get away with this but it would need some consultation as it's a regression on the current responsive behaviour.

> There are several things which might result in either the rulesStateStore or this target doc being written twice quickly...

I think it would be cleaner if we had a debounce so if this was called twice the second call just returned the original promise rather than chaining up another one. I'm worried about it being slow so the user spamming the refresh button which just makes it slower...

Reply:
> If I understand correctly this would mean if someone completed a task then the task completed count in the target wouldn't increment until midnight, right?

We write target emissions into the `_local/rulesStateStore` document. We write these updates every time they happen. We avoid conflicts using existing code (`stateChangeCallback`). Not writing to this causes the symtomps you have described (target counts don't increment in UI).

This function is writing to the `target-2019-11` document which replicates for analytics purposes. It doesn't control any functionality of the webapp - it is an analytics-only document.

Reply:
Ahh, right. Yes that's fine. Would you use some kind of scheduler so it runs every day, or only update the doc if it's been more than a day since the last update, or...?

Reply:
Only update the doc if it has been more than a day since the last update + the existing [freshness thread](https://github.com/medic/cht-core/pull/6114/files#diff-8d5be28dba4aeadf015d6ae8168834d6R52) that runs once per launch. How does that sound?

Reply:
Sounds good. My only concern is that the doc could get a little out of date but the chance of someone running the app without restarts long enough is low enough that that's unlikely to happen.